velox/expression/SignatureBinder.cpp

```diff
@@ -111,63 +118,134 @@ bool checkNamedRowField(
-  std::vector<Coercion> coercions;
-  return tryBind(false, coercions);
-}
+  const auto& varName = signature.baseName();
+  if (auto varIt = variables().find(varName); varIt != variables().end()) {
+    if (!varIt->second.isTypeParameter()) {
+      return true;
+    }
+
+    auto& varType = typeVariablesBindings_[varName];
+    if (!varType) {
+      if (!checkSignatureProperties(varIt->second, actualType)) {
+        return false;
+      }
+      varType = actualType;
+      return true;
+    }
+
+    if (TypeCoercer::coercible(actualType, varType)) {
+      return true;
+    }
+
+    if (TypeCoercer::coercible(varType, actualType)) {
+      if (!checkSignatureProperties(varIt->second, actualType)) {
+        return false;
+      }
+      varType = actualType;
+      return true;
+    }
+
+    return false;
+  }
+
+  const auto& signatureParams = signature.parameters();
+  const auto& actualParams = actualType->parameters();
 
-bool SignatureBinder::tryBind(
-    bool allowCoercions,
-    std::vector<Coercion>& coercions) {
-  if (allowCoercions) {
-    coercions.clear();
-    coercions.resize(actualTypes_.size());
+  const auto paramsCnt = std::min(signatureParams.size(), actualParams.size());
+  for (size_t i = 0; i < paramsCnt; ++i) {
+    if (actualParams[i].kind == TypeParameterKind::kType &&
+        !coercibleToTypeVars(signatureParams[i], actualParams[i].type)) {
+      return false;
+    }
   }
 
+  return true;
+}
+
+bool SignatureBinder::tryBind(std::vector<Coercion>* coercions) {
   const auto& formalArgs = signature_.argumentTypes();
   const auto formalArgsCnt = formalArgs.size();
 
   if (signature_.variableArity()) {
+    VELOX_DCHECK_GE(formalArgsCnt, 1);
     if (actualTypes_.size() < formalArgsCnt - 1) {
       return false;
     }
-
-    if (!isAny(signature_.argumentTypes().back())) {
-      if (actualTypes_.size() > formalArgsCnt) {
-        auto& type = actualTypes_[formalArgsCnt - 1];
-        for (auto i = formalArgsCnt; i < actualTypes_.size(); i++) {
-          if (!type->equivalent(*actualTypes_[i]) &&
-              actualTypes_[i]->kind() != TypeKind::UNKNOWN) {
-            return false;
-          }
+    if (!coercions && !isAny(signature_.argumentTypes().back()) &&
+        actualTypes_.size() > formalArgsCnt) {
+      const auto& variableType = actualTypes_[formalArgsCnt - 1];
+      if (!variableType) {
+        return false;
+      }
+      for (size_t i = formalArgsCnt; i < actualTypes_.size(); ++i) {
+        if (!actualTypes_[i]) {
+          return false;
+        }
+        if (actualTypes_[i] == UNKNOWN()) {
+          continue;
+        }
+        if (!actualTypes_[i]->equivalent(*variableType)) {
+          return false;
         }
       }
     }
-  } else {
-    if (formalArgsCnt != actualTypes_.size()) {
+  } else if (formalArgsCnt != actualTypes_.size()) {
+    return false;
+  }
+
+  // Phase 1: find least common var types
+  for (size_t i = 0; coercions && i < actualTypes_.size(); ++i) {
```

Comment:
This loop applies only if 'coercions' is not null, right? If so, it would be cleaner to put it inside an explicit 'if' statement.

```
if (coercions != nullptr) {
   for (...)
}
```

Reply:
ok

Reply:
Applied, see https://github.com/facebookincubator/velox/pull/15734/changes/861117750d189ac36413499955e1c28d07e1fb42